MdeModulePkg/Bus/Pci/PciHostBridgeDxe/PciHostResource.h

```diff
@@ -24,6 +24,12 @@ typedef enum {
   TypeMax
 } PCI_RESOURCE_TYPE;
 
+#if defined (MDE_CPU_IA32) || defined (MDE_CPU_X64)

```

Comment:
Need to avoid using #if in modules/libs.  Isn't there a better way to always loop through all resource types and for the CPU Arch/Si/Platform specific content to provide information if a specific resource type supported by the PCI standards is not supported by a specific CPU Arch/Si/Platform

Reply:
I can look into that and follow up.

Reply:
@Javagedes is planning to push an update to address this feedback.

Reply:
I had another idea on how to address this topic.  The PCI standard supports a number of BAR/resource types.  The PCI root bridge controllers typically support all of those BAR/resource types, but could potentially support a subset and be abstracted through the Si specific PCI Host Bridge Resource Allocation Protocol.  Some CPU Archs may not support all of the BAR/resource types that the PCI Root Bridge supports.

The implementation technique applied so far to address the CPU Arch subset is to adjust the start value of loop over a set of PCI_RESOURCE_TYPE enum values.  Given that different CPU Archs may have limitations in the middle of the enum range, it would be better if there was an internal CPU Arch specific API to check if a specific resource type is supported by that CPU Arch.

An X86 specific C file can provide an implementation that returns TRUE for all valid PCI_RESOURCE_TYPE values: 

```
BOOLEAN
IsPciResourceTypeSupported (
  PCI_RESOURCE_TYPE  PciResourceType
  )
{
  if (PciResourceType >= TypeMax) {
    return FALSE;
  }
  return TRUE;
}
```

For CPU Archs that do not support TypeIo, a CPU Arch specific file can provide the following implementation:

```
BOOLEAN
IsPciResourceTypeSupported (
  PCI_RESOURCE_TYPE  PciResourceType
  )
{
  if (PciResourceType >= TypeMax || PciResourceType == TypeIo) {
    return FALSE;
  }
  return TRUE;
}
```

Calls to this internal API can be made from the functions that loop over all the PCI_RESOURCE_TYPE values and can skip the ones that are not supported.

Reply:
Wouldn't it be better to simply check if the root bridge's I/O aperture `Base == MAX_UINT64` and/or `Limit == 0`?

This isn't really architecture-specific, since Arm can still support I/O BARs, but mapped into memory instead.